src/encoded/vis_defines.py

```diff
@@ -750,8 +747,42 @@ def lookup_token(self, token, dataset, a_file=None):
             if token.endswith("|multiple}"):
                 return "multiple biosamples"
             return "Unknown Biosample"
+        elif token == "{biosample_term_name}":
+            biosample_ontology = dataset.get('biosample_ontology')
+            if biosample_ontology is None:
+                return "Unknown Biosample"
+            if isinstance(biosample_ontology, dict):
+                return biosample_ontology['term_name']
+            if isinstance(biosample_ontology, list) and len(biosample_ontology) > 3:
+                return "Collection of %d Biosamples" % (len(biosample_ontology))
+            # The following got complicated because general Dataset objects
+            # cannot have biosample_ontology embedded properly. As a base class,
+            # some of the children, PublicationData, Project and 8 Series
+            # objects, have biosample_ontology embedded as array of objects,
+            # while experiment and annotation have it embedded as one single
+            # object. This becomes a problem when File object linkTo Dataset in
+            # general rather than one specific type. Current embedding system
+            # don't know how to map a property with type = ["array", "string"]
+            # in elasticsearch. Therefore, it is possible the
+            # "biosample_ontology" we got here is @id which should be embedded
+            # with the following code.
+            if not isinstance(biosample_ontology, list):
+                assert isinstance(biosample_ontology, str)
```

Comment:
Same here,

Is there a reason we're using assert instead of logging the exception?  Asserts are generally for crashing the run when something bad happens, not for production code.

...
We're making sure its not a list and then requiring it to be str.  Why not...
```
            if isinstance(biosample_ontology, str):
                biosample_ontology = [biosample_ontology]
                term_names = [
                    self.request.embed(type_obj_id, '@@object')['term_name']
                    for type_obj_id in biosample_ontology
                ]
                if len(term_names) == 1:
                    return term_names[0]
                else:
                    return term_names
             else: 
                 log.error("?")
                 return ??
```